modules/core/CameraThinLens.cpp

```diff
@@ -0,0 +1,33 @@
+#include "CameraThinLens.h"
+#include "Ray.h"
+#include "random.h"
+#include "Sampler.h"
+#include "macroses.h"
+
+namespace rt
+{
+	void CCameraThinLens::InitRay(Ray& ray, int x, int y, const Vec2f& sample)
+	{
+		CCameraPerspective::InitRay(ray, x, y, sample);
+
+		float lensr = getlensRadius();
+		float nBlades = getnBlades();
+		if (lensr > 0) {
+			// Sample point on lens
+			Vec2f pLens;
+			Vec2f sample(random::U<float>(), random::U<float>());
+			
+			//sample from uniformly distributed points in a regular polygon
+			RT_ASSERT(nBlades == 0 || (nBlades >= 3 && nBlades <= 16));
+			pLens = lensr * CSampler::uniformSampleRegularNgon(sample, nBlades, random::u<int>(1, nBlades));
```

Comment:
I think, that it makes sense to use an independent sampler here. Thus the number of samples for sampling a pixel may be independent from the number of samples used to sample a lens.
Let us add to the constrictor one more argument (like in constructor of an area light source)
`..., ptr_sampler_t pSampler = std::make_shared<CSamplerStratified>(4, true), ...` and this sampler will be used in this line.
I think that might reduce the amount of noise on distant out-of-focus objects...